packages/dds/tree/src/feature-libraries/modular-schema/discrepancies.ts

```diff
@@ -346,47 +288,68 @@ function trackObjectNodeDiscrepancies(
 	 * Then, the stored schema is iterated to find the third type.
 	 */
 
-	for (const [fieldKey, fieldStoredSchema] of view.objectNodeFields) {
-		viewFieldKeys.add(fieldKey);
-		if (
-			!stored.objectNodeFields.has(fieldKey) &&
-			fieldStoredSchema.kind !== storedEmptyFieldSchema.kind
-		) {
-			differences.push({
-				identifier: fieldKey,
-				mismatch: "fieldKind",
-				view: fieldStoredSchema.kind,
-				stored: storedEmptyFieldSchema.kind,
-			} satisfies FieldKindDiscrepancy);
-		} else {
-			differences.push(
-				...trackFieldDiscrepancies(
-					view.objectNodeFields.get(fieldKey) as TreeFieldStoredSchema,
-					stored.objectNodeFields.get(fieldKey) as TreeFieldStoredSchema,
-					fieldKey,
-				),
-			);
+	for (const result of compareMaps(view.objectNodeFields, stored.objectNodeFields)) {
+		const fieldKey = result.key;
+		switch (result.type) {
+			case "aExtra": {
+				if (result.value.kind === storedEmptyFieldSchema.kind) {
+					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
+					// (by way of omission). We treat these identically anyway.
+					break;
```

Comment:
do we leverage this distinction in some way? just wondering why it exists (didn't really know Forbidden was actually used)

Reply:
There's a fair amount of existing code which normalizes things to use appropriate forbidden terminology. One example is that trees use `storedEmptyFieldSchema` before they are initialized, which represents trees that cannot have any content and uses forbidden to do so:

![image](https://github.com/user-attachments/assets/04f5d2c3-5e35-49e8-b0b5-04155a10eb1c)

I found it convenient to continue using that pattern to handle compatibility checks. Feels similar to normalizing things to use explicit vs. implicit undefined.

AFAIK we don't expose this concept publicly directly, though it is possible for people to declare fields which are equivalent to forbidden (e.g. required field containing a tree which is unconstructable)

Reply:
Clarification for posterity as above comment isn't really precise: the screenshot of the root field schema represents a field which cannot have content (the forbidden field kind explicitly denotes that), whereas a required field containing no allowed types is actually a never field (it can never be satisfied: it requires content but no such content can exist). The last paragraph should use an optional field containing no allowed types or only unconstructable allowed types as its example.